pkg/operator/ceph/csi/operator.go

```diff
@@ -0,0 +1,155 @@
+	"github.com/pkg/errors"
+	cephv1 "github.com/rook/rook/pkg/apis/ceph.rook.io/v1"
+	cephclient "github.com/rook/rook/pkg/daemon/ceph/client"
+	v1 "k8s.io/api/core/v1"
+	kerrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/util/sets"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+func (r *ReconcileCSI) createCSIOperator(csiClusterConfig *cephclient.ClusterInfo, cluster cephv1.CephCluster) error {
+
+	logger.Info("Creating ceph-CSI operator")
+	err := r.createCSIOpCephCluster(csiClusterConfig, cluster)
+	if err != nil {
+		return errors.Wrap(err, "Failed to create ceph-CSI operator cephCluster CR")
+	}
+
+	err = r.createCSIOpCephCSIConfCR(cluster)
+	if err != nil {
+		return errors.Wrap(err, "Failed to create ceph-CSI operator config CR")
+	}
+
+	logger.Info("Successfully created ceph-CSI operator CR's")
+	return nil
+}
+
+func (r *ReconcileCSI) createCSIOpCephCluster(csiClusterConfig *cephclient.ClusterInfo, cluster cephv1.CephCluster) error {
+	logger.Info("Creating ceph-CSI operator ceph cluster CR")
+
+	csiCephCluster := &csiopv1a1.CephCluster{}
+	csiCephCluster.Name = cluster.Name
+	csiCephCluster.Namespace = cluster.Namespace
+	if cluster.Spec.CSI.ReadAffinity.Enabled {
+		csiCephCluster.Spec.ReadAffinity.CrushLocationLabels = cluster.Spec.CSI.ReadAffinity.CrushLocationLabels
+	}
+
+	mons := sets.New[string]()
+	for _, mon := range csiClusterConfig.Monitors {
+		mons.Insert(mon.Endpoint)
+	}
+	csiCephCluster.Spec.Monitors = sets.List(mons)
+
+	err := r.client.Create(r.opManagerContext, csiCephCluster)
+	if err != nil {
+		if kerrors.IsAlreadyExists(err) {
+			err = r.client.Update(r.opManagerContext, csiCephCluster)
+			if err != nil {
+				return errors.Wrapf(err, "failed to update CSI-operator ceph cluster CR %s", csiCephCluster.Name)
+			}
+		}
+		return errors.Wrapf(err, "failed to create CSI-operator ceph cluster CR %s", csiCephCluster.Name)
+	}
+
+	logger.Info("Successfully created ceph-CSI operator ceph cluster CR")
+	return nil
+}
+
+func (r *ReconcileCSI) createCSIOpCephCSIConfCR(cluster cephv1.CephCluster) error {
+
+	cephRadosNamespaceList := &cephv1.CephBlockPoolRadosNamespaceList{}
+	err := r.client.List(r.opManagerContext, cephRadosNamespaceList, &client.ListOptions{})
+	if err != nil {
+		return err
+	}
+
+	if len(cephRadosNamespaceList.Items) == 0 {
+		logger.Debug("no cephRadosNamespace is found")
+	} else {
+		for _, radosns := range cephRadosNamespaceList.Items {
+			if !radosns.DeletionTimestamp.IsZero() {
+				logger.Debugf("ceph radosnamespace %q is being deleting", cluster.Namespace)
+				return nil
+			}
+
+			csiOpConfig := &csiopv1a1.Config{}
+			csiOpConfig.Name = csiOPconfigCRName(radosns.Name)
+			csiOpConfig.Spec = csiopv1a1.ConfigSpec{
+				CephClusterRef: v1.LocalObjectReference{
+					Name: cluster.Name,
+				},
+				Rbd: &csiopv1a1.RbdConfigSpec{
+					RadosNamespace: radosns.Name,
+				},
+			}
+
+			err = r.client.Create(r.opManagerContext, csiOpConfig)
```

Comment:
Shall we create the rados namespace csi config CR during the CephBlockPoolRadosNamespace reconcile? Then whenever there is an update to those CRs, we would update the corresponding CSI config.